cloudinit/distros/freebsd.py

```diff
@@ -192,5 +194,40 @@ def update_package_sources(self):
             freq=PER_INSTANCE,
         )
 
+    @lru_cache()
+    def is_container(self) -> bool:
+        """return whether we're running in a container.
+        Cached, because it's unlikely to change."""
+        jailed, _ = subp.subp(["sysctl", "-n", "security.jail.jailed"])
+        if jailed.strip() == "0":
+            return False
+        return True
+
+    @lru_cache()
+    def virtual(self) -> str:
+        """return the kind of virtualisation system we're running under.
+        Cached, because it's unlikely to change."""
+        if self.is_container():
+            return "jail"
+        # map FreeBSD's kern.vm_guest to systemd-detect-virt, just like we do
+        # in ds-identify
+        VM_GUEST_TO_SYSTEMD = {
+            "hv": "microsoft",
+            "vbox": "oracle",
+            "generic": "vm-other",
+        }
+        vm, _ = subp.subp(["sysctl", "-n", "kern.vm_guest"])
+        vm = vm.strip()
+        if vm in VM_GUEST_TO_SYSTEMD:
+            return VM_GUEST_TO_SYSTEMD[vm]
+        return vm
+
+    @property
+    def is_virtual(self) -> Optional[bool]:
+        """Detect if running on a virtual machine or bare metal.
 
-# vi: ts=4 expandtab
+        This can fail on some platforms, so the signature is Optional[bool]
```

Comment:
How would this fail in a way that returns other than `bool`?

Reply:
If referring to the possibility of thrown exceptions, that shouldn't be part of the return type annotation.

Reply:
it can fail on Linux.
as in, it can fail to detect what type of platform its on. we're making the return type the same as the base implementation, which is Linux 

but i haven't observed this behaviour on FreeBSD